microservices-connector/api/v1alpha3/validating_webhook_test.go

```diff
@@ -0,0 +1,374 @@
+				map[string]Router{},
+				"a root node is required"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := validateRootExistance(tt.args.nodes, tt.args.fldPath); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("validateRootExistance() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_getKeys(t *testing.T) {
+	type args struct {
+		m map[string]Router
+	}
+	tests := []struct {
+		name string
+		args args
+		want []string
+	}{
+		{
+			name: "succes",
+			args: args{
+				m: map[string]Router{
+					"root": {},
+				},
+			},
+			want: []string{"root"},
+		},
+		{
+			name: "empty map",
+			args: args{
+				m: map[string]Router{},
+			},
+			want: []string{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := getKeys(tt.args.m); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getKeys() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_validateNames(t *testing.T) {
+	var errs field.ErrorList
+	type args struct {
+		nodes   map[string]Router
+		fldPath *field.Path
+	}
+	tests := []struct {
+		name string
+		args args
+		want field.ErrorList
+	}{
+		{
+			name: "duplicate service name",
+			args: args{
+				nodes: map[string]Router{
+					"root": {
+						Steps: []Step{
+							{
+								StepName: "Embedding",
+								Executor: Executor{
+									InternalService: GMCTarget{
+										ServiceName: "embedding_svc",
+									},
+								},
+							},
+							{
+								StepName: "Embedding",
+								Executor: Executor{
+									InternalService: GMCTarget{
+										ServiceName: "embedding_svc",
+									},
+								},
+							},
+						},
+					},
+				},
+				fldPath: field.NewPath("spec").Child("nodes"),
+			},
+			want: append(errs, field.Invalid(
+				field.NewPath("spec").Child("nodes").Child("root").Child("internalService").Child("serviceName"),
+				Step{
+					StepName: "Embedding",
+					Executor: Executor{
+						InternalService: GMCTarget{
+							ServiceName: "embedding_svc",
+						},
+					},
+				},
+				"service name: embedding_svc in node root already exists")),
+		},
+		{
+			name: "no error",
+			args: args{
+				nodes: map[string]Router{
+					"root": {
+						Steps: []Step{
+							{
+								StepName: "Embedding",
+								Executor: Executor{
+									InternalService: GMCTarget{
+										ServiceName: "embedding_svc1",
+									},
+								},
+							},
+							{
+								StepName: "Embedding",
+								Executor: Executor{
+									InternalService: GMCTarget{
+										ServiceName: "embedding_svc2",
+									},
+								},
+							},
+						},
+					},
+				},
+				fldPath: field.NewPath("spec").Child("nodes"),
+			},
+			want: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := validateNames(tt.args.nodes, tt.args.fldPath); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("validateNames() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```

Comment:
Can we add a use case for ENDPOINT? 
for example, 
```
      - name: Embedding
        internalService:
          serviceName: embedding-svc
          config:
            endpoint: /v1/embeddings
            TEI_EMBEDDING_ENDPOINT: tei-embedding-svc
```
in this step,  TEI_EMBEDDING_ENDPOINT: tei-embedding-svc means there must be a step with service name "tei-embedding-svc" in this node.

Reply:
Agree this is a good validation case

Reply:
Sure, I try to implement it if time permits.